aws/rds_test.go

```diff
@@ -73,17 +73,19 @@ func TestListRDS(t *testing.T) {
 	rdsName := "cloud-nuke-test" + util.UniqueID()
 	createTestRDSInstance(t, session, rdsName)
 
-	eds, err := getAllRdsInstances(session, time.Now())
+	rds, err := getAllRdsInstances(session, time.Now().Add(1*time.Hour))
 
 	if err != nil {
 		assert.Failf(t, "Unable to fetch list of RDS DB Instances", errors.WithStackTrace(err).Error())
 	}
 
-	assert.Contains(t, awsgo.StringValueSlice(eds), strings.ToLower(rdsName))
+	assert.Contains(t, awsgo.StringValueSlice(rds), strings.ToLower(rdsName))
 
-	nukeAllRdsInstances(session, eds)
+	defer func() {
+		nukeAllRdsInstances(session, rds)
 
-	rdsNames, err := getAllRdsInstances(session, time.Now().Add(1*time.Hour*-1))
+		rdsNames, _ := getAllRdsInstances(session, time.Now().Add(1*time.Hour))
 
-	assert.NotContains(t, awsgo.StringValueSlice(rdsNames), strings.ToLower(rdsName))
+		assert.NotContains(t, awsgo.StringValueSlice(rdsNames), strings.ToLower(rdsName))
+	}()
```

Comment:
The `defer` should actually happen earlier, right after the instance is created. I suspect you had trouble getting it to work because you want the reference to `rds`, which is returned by `getAllRdsInstances`.

To explain, `defer` is a golang keyword to get things to happen at  the end of the function, right before it exits. This works even if the function exits early due to a `t.Fatal` or `return` call. However, it is only scheduled when the execution pass reaches the `defer` routine. Since we always want to attempt to clean up the RDS instance we created, regardless of whether or not the `getAllRdsInstances` function is working, we want to schedule the routine immediately after the create call. But that means you need the `rds` variable to be set (for `nukeAllRdsInstances`) before the call to `getAllRdsInstances`. So how do you do that?

My suggestion would be to decouple the `nukeAllRdsInstances` call from `getAllRdsInstances`. That is, you want to call `nuke` on the RDS instance that you created, regardless of what `get` returns. So instead of nuking what is returned by get, you want the `create` call to return the rds instance it created and pass that to `nuke` in the `defer` call.

Does that make sense?

Reply:
Actually it doesn't make much sense to me, can we make a call?